external/pkg/docker-exec-ws/shellclient.go

```diff
@@ -0,0 +1,353 @@
+package dockerExecWS
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/taskcluster-cli/external/pkg/iopipes"
+	"github.com/taskcluster/taskcluster-worker/plugins/interactive/shellconsts"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+)
+
+const (
+	// MaxOutstandingBytes is the maximum number of unsent bytes in the stdin buffer.
+	MaxOutstandingBytes = 8 * 1024 * 1024
+	pingInterval        = 15 * time.Second
+	writeTimeout        = 2 * pingInterval
+	readTimeout         = 3 * pingInterval
+)
+
+var dialer = websocket.Dialer{
+	HandshakeTimeout: shellconsts.ShellHandshakeTimeout,
+	ReadBufferSize:   shellconsts.ShellMaxMessageSize,
+	WriteBufferSize:  shellconsts.ShellMaxMessageSize,
+}
+
+// ShellClient is the implementation of
+// github.com/taskcluster/taskcluster-worker/engines.Shell for v1 shells.
+type ShellClient struct {
+	ws       *websocket.Conn
+	done     chan struct{}
+	mWrite   sync.Mutex
+	cPause   *sync.Cond
+	paused   bool
+	complete atomics.Once
+	success  bool
+	err      error
+
+	stdin        io.WriteCloser
+	stdinReader  io.ReadCloser
+	stdout       io.ReadCloser
+	stdoutWriter *iopipes.DrainingPipeWriter
+	stderr       io.ReadCloser
+	stderrWriter *iopipes.DrainingPipeWriter
+}
+
+// Dial will built a proper shell websocket URL, connect to the server and return an initialized shell client.
+func Dial(socketURL string, command []string, tty bool) (*ShellClient, error) {
+	u, err := url.Parse(socketURL)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid socketURL: %s, parsing error: %s",
+			socketURL, err)
+	}
+	q := u.Query()
+
+	// Ensure the URL has ws or wss as scheme
+	switch u.Scheme {
+	case "http":
+		u.Scheme = "ws"
+	case "https":
+		u.Scheme = "wss"
+	} // Set command arguments overwriting any existing querystring values
+	q.Del("command")
+	if len(command) > 0 {
```

Comment:
you don't need this "if", do you... the loop will just be a no-op if there is zero...

Reply:
done